### PR TITLE
Add 4.0 tests, remove hardcoded cluster names from tests [K8SSAND-890] [K8SSAND-862] 

### LINKED ISSUE #112: K8SSAND-862 ⁃ Refactor e2e tests so they can be run against clusters other than ones created by setup script
We should be able to run the e2e tests against any clusters provided that they have sufficient resources and pods have routable IPs. There are at least a couple things that need to be fixed to support this.

The k8s context names are hard coded in a bunch of places. Consider the following example:

```go
t.Log("check that datacenter dc1 is ready")
dcKey := framework.ClusterKey{K8sContext: "kind-k8ssandra-0", NamespacedName: types.NamespacedName{Namespace: namespace, Name: "dc1"}}
withDatacenter := f.NewWithDatacenter(ctx, dcKey)
```

`kind-k8ssandra-0` corresponds to what is generated by `setup-kind-multicluster.sh`. The contexts need to be configurable.

The second issue involves the kubeconfig files used by the tests. When you create a kind cluster the kubeconfig entry that is creates uses localhost for the api server address. We generate a second kubeconfig that uses the docker network address. There are some checks that make sure each of these kubeconfig files exists. If I want to run the tests against GKE or EKS clusters for instance, these extra steps shouldn't be necessary.

When these issues are resolved we can start working on setting some GHA workflows for running the tests with the various cloud providers.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-862) by [Unito](https://www.unito.io)
┆Reviewer: 5e66ca0cae59460cfb01a13c
┆fixVersions: k8ssandra-operator-v1.0.1
┆friendlyId: K8SSAND-862
┆priority: Medium

### LINKED ISSUE #130: K8SSAND-890 ⁃ #130 - Update e2e tests in GHA to against both Cassandra 3.11.x and 4.0.x
In GHA we only run the e2e tests against Cassandra 3.11.x right. We should be running the tests against 4.0.x as well. If we already had that set up, we would have caught #129 earlier.

In the e2e tests we configure and deploy the K8ssandraCluster with kubectl. We already have support in place for passing arguments to the tests. In the Makefile we have:

```
PHONY: e2e-test
e2e-test:
ifdef E2E_TEST
	@echo Running e2e test $(E2E_TEST)
	go test -v -timeout 3600s ./test/e2e/... -run="$(E2E_TEST)" -args $(TEST_ARGS)
else
	@echo Running e2e tests
	go test -v -timeout 3600s $(TEST_ARGS) ./test/e2e/...
endif
```

Note the `-args` We can pass a `cassandraVersion` argument here.  Then we can create a separate K8ssandraCluster manifest for 4.0.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-890) by [Unito](https://www.unito.io)
┆Reviewer: 5e66ca0cae59460cfb01a13c
┆fixVersions: k8ssandra-operator-v1.0.0
┆friendlyId: K8SSAND-890
┆priority: Medium

### PR DESCRIPTION
**Pull request by burmanm**
**What this PR does**:
Substitutes certain parameters, such as cassandraVersion, controlPlane and dataPlane clusters to avoid hardcoded values. Adds tests against different versions of Cassandra also.

**Which issue(s) this PR fixes**:
Fixes #112
Fixes #130

**Checklist**
- [ ] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [ ] CHANGELOG.md updated (not required for documentation PRs)
- [x] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

**Comment by burmanm** [comment]
The tests seem quite flaky. I didn't find any issues running locally, but GHA fail to different timeouts..

```
/home/michael/bin/kubectl --context kind-k8ssandra-0 -n multi-dc-kqqcc6 apply -f -
    suite_test.go:484: check that the K8ssandraCluster was created
    suite_test.go:685: check that datacenter dc1 in cluster kind-k8ssandra-0 is ready
    suite_test.go:492: check k8ssandra cluster status
    suite_test.go:685: check that datacenter dc2 in cluster kind-k8ssandra-1 is ready
    suite_test.go:512: check k8ssandra cluster status
    suite_test.go:535: check that nodes in dc1 see nodes in dc2
/home/michael/bin/kubectl --context kind-k8ssandra-0 -n multi-dc-kqqcc6 exec -i test-dc1-rack1-sts-0 -c cassandra -- nodetool status
time="2021-09-16T08:49:20+03:00" level=info msg="nodetool status output" Count=6
    suite_test.go:543: check nodes in dc2 see nodes in dc1
/home/michael/bin/kubectl --context kind-k8ssandra-1 -n multi-dc-kqqcc6 exec -i test-dc2-rack1-sts-0 -c cassandra -- nodetool status
time="2021-09-16T08:49:26+03:00" level=info msg="nodetool status output" Count=6
time="2021-09-16T08:49:26+03:00" level=info msg="dumping cluster info"
time="2021-09-16T08:49:27+03:00" level=info msg="deleting all K8ssandraClusters" Namespace=multi-dc-kqqcc6
time="2021-09-16T08:49:27+03:00" level=info msg="deleting all Stargates" Namespace=multi-dc-kqqcc6
time="2021-09-16T08:49:37+03:00" level=info msg="deleting all CassandraDatacenters" Namespace=multi-dc-kqqcc6
time="2021-09-16T08:49:57+03:00" level=info msg="deleting all ReplicatedSecrets" Namespace=multi-dc-kqqcc6
time="2021-09-16T08:50:07+03:00" level=info msg="deleting namespace" Context=kind-k8ssandra-1 Namespace=multi-dc-kqqcc6
time="2021-09-16T08:50:07+03:00" level=info msg="deleting namespace" Context=kind-k8ssandra-0 Namespace=multi-dc-kqqcc6
--- PASS: TestOperator (756.96s)
    --- PASS: TestOperator/CreateMultiDatacenterCluster (756.96s)
PASS
ok      github.com/k8ssandra/k8ssandra-operator/test/e2e        757.001s
```

**Comment by jsanda** [comment]
#148 should address the flakes 🤞 @adutra is working on the Stargate failures.

**Comment by burmanm** [comment]
I see the #148 getting more complex and people adding more and more new references to hardcoded cluster names, so it would be nice to merge this before rebase-hell ;)

**Comment by burmanm** [comment]
@jsanda ping pong.. this is getting painful to rebase every day (twice today)

**Comment by jsanda** [comment]
Sorry for the churn. Hopefully that is done now that #154 is merged.

**Comment by burmanm** [comment]
(the test failure is because of a problem in the main branch, not this branch)

File: test/e2e/suite_test.go [hunk 1]
```diff
@@ -209,14 +217,53 @@ func beforeTest(t *testing.T, namespace, fixtureDir string, f *framework.E2eFram
...
 
-	if err := kubectl.Apply(kubectl.Options{Namespace: namespace, Context: f.ControlPlaneContext}, fixtureDir); err != nil {
-		t.Log("kubectl apply failed")
+	// Substitute values in Kustomize templates for tests and apply them
+	fixtureFiles, err := ioutil.ReadDir(fixtureDir)
+	if err != nil {
 		return err
 	}
 
+	for _, filu := range fixtureFiles {
```
[the remark this answers is not in the document]
Comment (burmanm) [reply test/e2e/suite_test.go:226]: There is already an official version with the fix included.

[the remark this answers is not in the document]
Comment (jsanda) [reply test/e2e/suite_test.go:226]: It looks like 1.0.36 was released ~15 hours ago. I don't see the new image tag on Docker Hub yet so for the moment, I think we still need the change. Regardless of whether a new image is available, I still think it is easier to maintain a separate set of 4.0 manifests.

**Comment by jdonenine** [comment]
Closing this PR as it's now very stale, but the content is very much desired and now outlined in issue #360 to be added back post-1.0.0

**Comment by jdonenine** [comment]
The changes originally made here are also relevant to https://github.com/k8ssandra/k8ssandra-operator/issues/112

**Pull request closed without merging**